### PR TITLE
Expose network stats for pods

### LINKED ISSUE #368: Heapster not getting network rx and tx bytes info
Hi

I am running heapster on kubernetes. Several pods are running but all of them giving 0 values in rx and tx network stats.
On doing docker stats <container-id>, I found that network stats are also coming 0 for that container but network stats are coming in cadvisor correctly.
Network stats for the pod gcr.io/google_containers/pause:0.8.0 are coming correctly but they are not communicated into the influxdb by the heapster. The containers corresponding to this gcr.io/google_containers/pause:0.8.0 are showing 0 network stats. 

What should I do to get correct network stats ?
Please help !

### PR DESCRIPTION
**Pull request by jimmidyson**
Fixes #368 

Dirty: using leaky infra container name, regex for container name to pod mapping, but seems to work OK...

/cc @vishh @akash010 @smarterclayton @simon3z

### CODE REVIEW

**Comment by smarterclayton** [comment]
If leaky is imported just to get pod infra container name, it's not worth it.  I'd just have it be a constant or parameter in this code.

**Comment by jimmidyson** [comment]
retest this please

**Comment by jimmidyson** [comment]
Removed `leaky` package dependency.

**Comment by jimmidyson** [comment]
Got more work to do on this to expose via the API so changing to WIP - please don't merge.

@vishh @mwielgus Do you know if the `Jenkins GCE e2e` should pass? Seeing unrelated flake in logs:

    No error is expected but got expected [gcm] sinks, found []

**Comment by mwielgus** [comment]
Yeah, the problem seems to be unrelated. BTW, we will have to redo this work in heapster-scalability branch.

**Comment by jimmidyson** [comment]
@mwielgus Are you targeting the new metrics APIs in the scalability branch? Sorry I've not kept up to speed with it.

**Comment by mwielgus** [comment]
If the new API is delivered on time then yes we will switch, if not we will keep using the old one.

**Comment by jimmidyson** [comment]
Ready for review please. Jenkins GCE e2e is flaky & needs to either disabled or fixed separately to this PR.

File: api/v1/types/model_types.go [hunk 1]
```diff
@@ -50,10 +50,12 @@ type StatsResponse struct {
...
 }
 
-// An ExternalEntityListEntry represents the latest CPU and Memory usage of a model entity.
+// An ExternalEntityListEntry represents the latest CPU, Memory & network usage of a model entity.
 // A model entity can be a Pod, a Container, a Namespace or a Node.
 type ExternalEntityListEntry struct {
 	Name     string `json:"name"`
 	CPUUsage uint64 `json:"cpuUsage"`
 	MemUsage uint64 `json:"memUsage"`
+	RxBytes  uint64 `json:"rx_bytes"`
```
Comment (vishh) [inline api/v1/types/model_types.go]: Do we need network metrics in the model or is it necessary only for monitoring purposes?
The cost for adding metrics to the model is kindda high for now.

File: sources/datasource/kubelet.go [hunk 2]
```diff
@@ -176,11 +185,23 @@ func (self *kubeletSource) getAllContainers(url string, start, end time.Time) ([
 	}
 
 	// TODO(vmarmol): Use this for all stats gathering.
-	// Don't include the Kubernetes containers, those are included elsewhere.
+	// Only include the Kubernetes infra containers, the others are collected via the kubelet pod stats API.
```
Comment (vishh) [inline sources/datasource/kubelet.go]: I think we want to collect the system containers like `/kubelet`, `/docker-daemon`, etc which should have been the existing behavior.

File: api/v1/types/model_types.go (hunk 1 above)
Comment (jimmidyson) [inline api/v1/types/model_types.go]: So are you saying it shouldn't be returned by API queries, only passed to sinks? I'd prefer it to be exposed via the REST API & that means using the model, doesn't it?

File: sources/datasource/kubelet.go (hunk 2 above)
Comment (jimmidyson) [inline sources/datasource/kubelet.go]: Sorry that's a *bad* comment! Will update - raw containers are of course retained.

File: api/v1/types/model_types.go (hunk 1 above)
Comment (vishh) [inline api/v1/types/model_types.go]: Yeah. Is it **required** to be exposed via APIs as well?

Comment (jimmidyson) [inline api/v1/types/model_types.go]: Point me to the requirements :) Don't we make it up as we go along? ;)

Honestly, if it's too expensive to store a couple of extra values per pod then I don't mind dropping it from there & just leave it as passed to sinks.

File: sources/datasource/kubelet.go (hunk 2 above)
Comment (jimmidyson) [inline sources/datasource/kubelet.go]: Updated

File: api/v1/types/model_types.go (hunk 1 above)
Comment (vishh) [inline api/v1/types/model_types.go]: I'd personally prefer exposing this and filesystem stats as well. Its just that the default resource limits in Kube has to change once these metrics are added. So as long as we can go fix the limits, then adding these metrics is OK by me :)

Comment (jimmidyson) [inline api/v1/types/model_types.go]: OK how about I remove from the model for this version & revisit it for @mwielgus' rewritten version?

Comment (vishh) [inline api/v1/types/model_types.go]: SGTM. So is monitoring the primary use case?

Comment (jimmidyson) [inline api/v1/types/model_types.go]: Monitoring & accounting, yes.

Comment (jimmidyson) [inline api/v1/types/model_types.go]: Can't think of a use for autoscaling on network traffic right now.

File: manager/manager.go [hunk 3]
```diff
@@ -187,6 +191,44 @@ func (rm *realManager) housekeep(start, end time.Time) {
...
+			if !podNameFound || !podNamespaceFound {
+				continue
+			}
+
+			// Find a pod with these details.
+			for podIndex := range sd.data.Pods {
+				pod := &sd.data.Pods[podIndex]
+				// If we find a matching pod then add the container to the pod's containers slice.
+				if pod.Name == podName && pod.Namespace == podNamespace {
+					cont.Hostname = pod.Hostname
```
Comment (vishh) [inline manager/manager.go]: Ideally, the pod infra container should be hidden inside heapster. That way if we collect metrics from rocket for example, which does not need an infra container, we will not break users of heapster.

File: sources/datasource/kubelet.go [hunk 4]
```diff
@@ -32,7 +33,12 @@ import (
 )
 
 // TODO(vmarmol): Use Kubernetes' if we export it as an API.
-const kubernetesPodNameLabel = "io.kubernetes.pod.name"
+// Copied from k8s.io/kubernetes/pkg/kubelet/dockertools/labels.go - not exported there :(
```
Comment (vishh) [inline sources/datasource/kubelet.go]: The reason is that these labels are not intended to be exposed :)

**Comment by vishh** [comment]
General structure LGTM. Thanks @jimmidyson !!

File: manager/manager.go (hunk 3 above)
Comment (jimmidyson) [inline manager/manager.go]: Suggestions on how to do that? Right now I can't think of one tbh.

**Comment by jimmidyson** [comment]
Network stats are now only sent to sinks, not retrievable via Heapster API as discussed with @vishh.

File: manager/manager.go (hunk 3 above)
Comment (vishh) [inline manager/manager.go]: We need stats at the pod level in addition to container level.

Comment (jimmidyson) [inline manager/manager.go]: Bit confused by this. We're adding the infra container to the appropriate pod so this is done.

**Comment by jimmidyson** [comment]
@vishh Please can you let me know what is outstanding for this PR to be merged?

**Comment by vishh** [comment]
The only issue with this PR is that of exposing the infra container. I'd prefer adding pod level metrics and exposing network as a pod level metrics.

**Comment by jimmidyson** [comment]
This is how's it's been done in `heapster-scalability` branch & I'm not going to duplicate that functionality as this is just a quick fix until `heapster-scalability` branch becomes master.

**Comment by vishh** [comment]
Ok then. No issues in that case.

**Comment by jimmidyson** [comment]
Thanks @vishh.

Merging.

**Pull request merged**